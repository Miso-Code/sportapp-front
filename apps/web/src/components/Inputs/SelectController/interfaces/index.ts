import { SelectProps, FormControlProps, InputLabelProps } from '@mui/material'
import { Control, FieldValues, Path } from 'react-hook-form'

export interface Props<T extends FieldValues = FieldValues> {
	control: Control<T> | undefined
	name: Path<T>
	label: string
	options: Options[]
	selectProps?: SelectProps
	formControlProps?: FormControlProps
	inputLabelProps?: InputLabelProps
	isTranslated?: boolean
	isDisabled?: boolean
}

export interface Options {
	value: string
	label: string
}
