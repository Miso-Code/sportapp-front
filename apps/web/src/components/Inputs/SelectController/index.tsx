import {
	FormControl,
	FormHelperText,
	InputLabel,
	MenuItem,
	Select
} from '@mui/material'
import { Props } from 'components/Inputs/SelectController/interfaces'
import { Controller, FieldValues } from 'react-hook-form'
import { useTranslation } from 'react-i18next'

export default function SelectController<T extends FieldValues = FieldValues>({
	control,
	name,
	label,
	options,
	formControlProps,
	inputLabelProps,
	selectProps
}: Props<T>) {
	const { t } = useTranslation()
	return (
		<>
			<Controller
				control={control}
				name={name}
				render={({
					field: { onChange, value, ...field },
					fieldState: { error }
				}) => (
					<>
						<FormControl
							fullWidth
							error={!!error}
							{...formControlProps}>
							<InputLabel
								id={`${name}-select-label`}
								{...inputLabelProps}>
								{label}
							</InputLabel>
							<Select
								{...selectProps}
								labelId={`${name}-select-label`}
								id={name}
								value={value}
								label={label}
								onChange={onChange}
								{...field}>
								{options.map((option) => (
									<MenuItem
										key={option.value}
										value={option.value}>
										{t(option.label)}
									</MenuItem>
								))}
							</Select>
							<FormHelperText>
								{error?.message ? t(error.message) : ''}
							</FormHelperText>
						</FormControl>
					</>
				)}
			/>
		</>
	)
}
