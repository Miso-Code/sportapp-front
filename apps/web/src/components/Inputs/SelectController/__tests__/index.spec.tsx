import SelectController from 'components/Inputs/SelectController'
import { render, RenderResult, renderHook } from '@testing-library/react'
import { useForm } from 'react-hook-form'

describe('SelectController', () => {
	let wrapper: RenderResult

	beforeEach(() => {
		const { result } = renderHook(() => useForm())
		wrapper = render(
			<SelectController
				control={result.current.control}
				name='select'
				label='Select'
				options={[
					{ label: 'Option 1', value: '1' },
					{ label: 'Option 2', value: '2' }
				]}
			/>
		)
	})

	afterEach(() => {
		wrapper.unmount()
	})

	it('should render the component', () => {
		expect(wrapper.container).toMatchSnapshot()
	})

	it('should display helper text when there is an error in input', () => {
		const { result } = renderHook(() => useForm())
		result.current.setError('select', {
			type: 'manual',
			message: 'Invalid Select'
		})
		wrapper = render(
			<SelectController
				control={result.current.control}
				name='select'
				label='Select'
				selectProps={{ fullWidth: true, defaultValue: '1' }}
				options={[
					{ label: 'Option 1', value: '1' },
					{ label: 'Option 2', value: '2' }
				]}
			/>
		)

		expect(wrapper.getByText('Invalid Select')).toBeInTheDocument()
	})
})
