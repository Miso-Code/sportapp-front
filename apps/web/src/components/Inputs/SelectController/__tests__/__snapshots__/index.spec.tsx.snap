// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectController should render the component 1`] = `
<div>
  <div
    class="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink="true"
      id="select-select-label"
    >
      Select
    </label>
    <div
      class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl  css-1ualgfl-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
    >
      <div
        aria-controls=":r0:"
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-labelledby="select-select-label mui-component-select-select"
        class="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
        id="mui-component-select-select"
        role="combobox"
        tabindex="0"
      >
        Option 1
      </div>
      <input
        aria-hidden="true"
        aria-invalid="false"
        class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
        id="select"
        name="select"
        tabindex="-1"
        value="1"
      />
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
        data-testid="ArrowDropDownIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
      <fieldset
        aria-hidden="true"
        class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="css-14lo706"
        >
          <span>
            Select
          </span>
        </legend>
      </fieldset>
    </div>
    <p
      class="MuiFormHelperText-root MuiFormHelperText-sizeMedium MuiFormHelperText-contained MuiFormHelperText-filled css-1wc848c-MuiFormHelperText-root"
    />
  </div>
</div>
`;
